README.md
Introduction
============

dcaenc is an open-source implementation of the DTS Coherent Acoustics codec.
Only the core part of the encoder is implemented, because the public
specification (ETSI TS 102 114 V1.2.1) is insufficient for implementing
anything else.

The DTS technology is heavily patented. Please do not download and do not use 
this software if you live in a country such as the USA where software patents are
legal and you don't have a patent license from DTS, Inc.

The project page is at http://aepatrakov.narod.ru/dcaenc/

The latest code can be obtained from git:

 git clone git://gitorious.org/dtsenc/dtsenc.git

Windows-specific notes
======================

There is a fork of dcaenc specially targetted at Windows users. You can get
it from http://gitorious.org/~mulder/dtsenc/mulders-dtsenc . As compared to
this the original version, that fork adds a Visual Studio project file, a
progress indicator and support for Unicode file names on Windows.
It also breaks Linux support.

ffdcaenc (Free Fast DTS Encoder) is a fork from mulders-dtsenc. It adds
support for 24 bit input files, multiple mono input files, flags for options
needed for DTS CD and DVD, and changes the core dcaenc library to support
strictly-compliant DVD bitrates as expected by AudioMuxer and Muxman.


The command-line application
============================

The ffdcaenc command line utility converts multichannel wav files to DTS (or
one or multiple mono files to DTS). The resulting DTS files can be written to an audio
CD and played via a digital connection (SPDIF or HDMI) to a receiver, or
used as sound tracks for a DVD.

Usage:

 ffdcaenc -i input.wav -o output.dts -b bitrate
 
For 5.1, the input wav file should have the same channel order as defined by SMPTE and ITU,
i.e.: left, right, center, lfe, surround left, surround right.
 
or

 ffdcaenc -m -0 left.wav -1 right.wav -2 center.wav -3 lfe.wav -4 ls.wav -5 rs.wav -o output.dts -b bitrate

using multiple mono wave files. The following mono input file combinations are supported:

	1.0		-2 center.wav
	1.1		-2 center.wav -3 lfe-wav
	2.0		-0 left.wav -1 right.wav
	2.1		-0 left.wav -1 right.wav -3 lfe.wav
	3.0		-0 left.wav -1 right.wav -2 center.wav
	3.1		-0 left.wav -1 right.wav -2 center.wav -3 lfe.wav
	4.0		-0 left.wav -1 right.wav -4 ls.wav -5 rs.wav
	4.1		-0 left.wav -1 right.wav -4 ls.wav -5 rs.wav -3 lfe.wav
	5.0		-0 left.wav -1 right.wav -2 center.wav -4 ls.wav -5 rs.wav
	5.1		-0 left.wav -1 right.wav -2 center.wav -4 ls.wav -5 rs.wav -3 lfe.wav

The files can be in any order on the command line.

Some destinations require a specific bitrate to be specified. To create a
CD-compatible DTS file from a multichannel 44.1kHz sampel rate file, run:

 ffdcaenc -e -r -i input.wav -o output.dts -b 1411.2
 
The dts file will need to be converterd to a wav file in order to be burned to a cd.
This can be done with spdifer, part of the ac3filter projects,
 http://ac3filter.net/

 spdifer output.dts output.wav -wav	

To create a DVD-compatible track from a multichannel (or 6 mono) wav file(s) that has a
48 kHz sample rate:

 ffdcaenc -i input.wav -o output.dts -b 1509.75

or for a half-rate output:

 ffdcaenc -i input.wav -o output.dts -b 754.5

The resulting dts file can be muxed with titles/still images using AudioMuxer,
	http://www.surroundbyus.com/sbu/viewtopic.php?f=7&t=135

or with video files using muxman
	http://stnsoft.com/Muxman/versions.shtml
	
or other tools that support muxing of dts streams.
  
Known bug: wav files with floating-point samples are misinterpreted as
containing 32-bit integer samples.

ORIGINAL DCAENC NOTES
=====================

The original dcaenc can be compiled for Windows, here is how.

The recommended compiler for dcaenc on Windows is MinGW. Get it from
http://sourceforge.net/projects/mingw/files/Installer/ . You may want to use
either the mingw-get.exe utility directly (just copy it to c:\mingw\bin,
install nothing else), or use a graphical installer. In the latter case,
install just the C compiler.

After installation, add c:\mingw\bin and c:\mingw\msys\1.0\bin to the Path
environment variable (right-click My Computer -> Properties -> Advanced ->
Environment Variables) if it is not already there. Then install the required
MinGW components from the command prompt:

 mingw-get install mingw32-base mingw32-autotools

For autotools to work, you also have to copy or rename the
c:\mingw\msys\1.0\etc\fstab.sample file to c:\mingw\msys\1.0\etc\fstab.

Installation
============

The dcaenc library has no external dependencies. The optional ALSA plugin
that comes with it, obviously, depends on the ALSA library compiled with
support for external plugins.

See the file INSTALL for the generic installation instructions.
Quick summary:

 autoreconf -f -i -v       # only if building from git
 ./configure --prefix=/usr # with any other prefix, ALSA won't find the plugin
 make
 su -c 'make install'      # Or, on Ubuntu: sudo make install

Package contents
================

The dcaenc package contains the libdcaenc.so shared library, the corresponding
dcaenc.h header, the dcaenc command line tool, and (optionally) the ALSA
plugin.

The libdcaenc library
=====================

The library may be useful if you want to create DTS streams in your own
application.

To do so, you have to call the following functions:

dcaenc_context dcaenc_create(
    int sample_rate,
    int channel_config,
    int approx_bitrate,
    int flags);

This function creates a library context according to its parameters. The
resulting context is an opaque value that should be passed to the other library
functions as the first parameter.

The sample rate must be one of the following values: 32000, 44100, 48000 or
those divided by 2 or 4. The channel configuration is one of the
DCAENC_CHANNELS_* defines listed in the dcaenc.h header. Note that values
greater than DCAENC_CHANNELS_3FRONT_2REAR always return an error now because
their encoding requires the Xch extension that is undocumented and thus not
implemented. Only non-LFE channels have to be specified here. The approximate
bitrate is specified in bits per second and may be rounded up by the library.

The flags parameter should be a logical OR of zero or more of the
DCAENC_FLAG_* defines. Their meanings:

DCAENC_FLAG_28BIT: use only 28 out of 32 bits in each four bytes of output.
This may be useful if the loudness of the hiss resulting from
mis-interpretation of the encoded stream as PCM must be reduced. However,
this also results in the reduction of the effective bitrate by 12.5%. DTS
CDs are usually encoded with this option.

DCAENC_FLAG_BIGENDIAN: produce a big-endian variant of the DTS bitstream.
This may be useful for writing the stream as a DVD sound track.

DCAENC_FLAG_LFE: indicates that a separate LFE channel has to be added to the
layout indicated by the channel_config parameter.

DCAENC_FLAG_PERFECT_QMF: selects a "perfect-reconstruction" version of the
quadrature mirror filter. This reduces distortions inherent in the filterbank
design, but makes the filter output more sensitive to quantization errors
introduced later in the encoder. libdca version 0.0.5 or earlier will not be
able to decode the resulting stream correctly due to a bug (mistyped table)
in it. It makes no sense to use this option for streams with bitrate of
0.3 Mbps per channel or less.

DCAENC_FLAG_IEC_WRAP: wraps DTS frames as defined by the IEC 61937-5 standard
for transmission over SPDIF. Namely, the library adds a 8-byte header to each
frame and pads the frame with zeroes to achieve the same bitrate as a stereo
16-bit PCM stream.

On any error, dcaenc_create() returns NULL. There is no way to find out the
reason for the error.

int dcaenc_bitrate(dcaenc_context c);

Returns the actual bitrate that is used by the library.

int dcaenc_input_size(dcaenc_context c);

Returns the size of the input buffer that your application has to submit, in
samples. Now this value is always equal to 512.

int dcaenc_output_size(dcaenc_context c);

Returns the size of the output buffer that the application should provide,
in bytes.

int dcaenc_convert_s32(
    dcaenc_context c,
    const int32_t *input,
    uint8_t *output);

Performs the conversion of PCM samples stored in the input buffer to the
DTS bitstream, stores one frame of the encoded bitstream in the output buffer.

NEWS
dcaenc-1: 2011-09-17 http://aepatrakov.narod.ru/index/0-2
* Initial release.

dcaenc-2: 2012-01-28 http://aepatrakov.narod.ru/index/0-2
* Minor cleanups
* Support for 5.0 channel encoding
* Fixed infinite loop on too-high bitrates
* Fixed buffer overflow on attempts to encode a 7.1 file

mulder2@gmx.de Windows port 2013-4-22	https://gitorious.org/~mulder/dtsenc/mulders-dtsenc

ffdcaenc-2.1: 2013-05-22 
* In command line tool; added support for 24 bit input files
* Changed PCM Source Resolution from 16 to 24 bits in DTS header
* In command line tool; added -r flag for reduced bit depth of output for DTS CDs
* In command line tool; added -m flag and support for 6 mono files input
* Changed Frame Size Calculation to allow for identical frame sizes as in commercial encoders
  Specifically -b 1509.75 now gives frame size acceptable to AudioMuxer and Muxman for DVD creation

ffdcaenc-2.1.1: 2013-05-31
* Bugfix for multi-mono (-m)
* In command line tool, allow for 5.0 in multi mono (5 inputs and no LFE)
	
ffdcaenc-2.1.2: 2013-06-03
* Generalize multi-mono to be as flexible as multi channel input

ffdcaenc-2.1.3: 2013-06-05
* Formalize ffdcaenc as a fork (documentation and command line target name)
* Further generalize multi-mono to be as flexible as multi channel input


main.c
/* 
 * This file is part of dcaenc.
 *
 * Copyright (c) 2008-2012 Alexander E. Patrakov <patrakov@gmail.com>
 *
 * dcaenc is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * dcaenc is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with dcaenc; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 *
 * Portions of this file changed by Alexei Andropov for ffdcaenc
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include "config.h"

#include "ffdcaenc.h"
#include "wavfile.h"
#include "xgetopt.h"
#include "compiler_info.h"

//extern const int32_t prototype_filter[512];

static char status[4] = {'|','/','-','\\'};
static const int32_t AUTO_SELECT = -1;

#define BUFFSIZE_SPL 512
#define BUFFSIZE_CHN 6
#define BUFFSIZE_OUT 16384

static int32_t dcaenc_main(int32_t argc, char *argv[])
{
	dcaenc_context c;
	int32_t data[BUFFSIZE_SPL * BUFFSIZE_CHN];
	uint8_t output[BUFFSIZE_OUT];
	wavfile * f;
	wavfile * mono_f[6];
	FILE * outfile;
	const char *error_msg;
	uint32_t samples_total;
	uint32_t mono_samples_total[6];
	size_t samples_read;
	size_t samples_read_total, current_pos;
	double percent_done;
	int32_t i;
	int32_t bitrate;
	int32_t wrote;
	int32_t counter;
	int32_t status_idx;
	int32_t show_ver;
	int32_t show_help;
	int32_t ignore_len;
	int32_t enc_flags;
	int32_t channel_config;
	int32_t has_lfe;
	int32_t multi_mono;
	int32_t any_input;
	int32_t mono_channels;
	int32_t an_input;
	xgetopt_t opt;
	char t;
	char *file_input;
	char *file_output;
	char *mono_file_input[6];

	static const int32_t channel_map[6] = {DCAENC_CHANNELS_MONO, DCAENC_CHANNELS_STEREO, DCAENC_CHANNELS_3FRONT,
		DCAENC_CHANNELS_2FRONT_2REAR, DCAENC_CHANNELS_3FRONT_2REAR, DCAENC_CHANNELS_3FRONT_2REAR};

	fprintf(stderr, " \n");

	// ----------------------------

	file_input = NULL;
	for (i=0; i<6;i++)
		mono_file_input[i] = NULL;

	bitrate = 0;
	enc_flags = DCAENC_FLAG_BIGENDIAN;
	channel_config = AUTO_SELECT;
	has_lfe = 0;
	multi_mono = 0;
	show_ver = 0;
	ignore_len = 0;
	show_help = 0;

	memset(&opt, 0, sizeof(xgetopt_t));
	while((t = xgetopt(argc, argv, "i:o:b:c:0:1:2:3:4:5:fhmlrev", &opt)) != EOF)
	{
		switch(t)
		{
		case 'i':
			file_input = opt.optarg;
			break;
		case '0':
			mono_file_input[0] = opt.optarg;
			break;
		case '1':
			mono_file_input[1] = opt.optarg;
			break;
		case '2':
			mono_file_input[2] = opt.optarg;
			break;
		case '3':
			mono_file_input[3] = opt.optarg;
			break;
		case '4':
			mono_file_input[4] = opt.optarg;
			break;
		case '5':
			mono_file_input[5] = opt.optarg;
			break;
		case 'o':
			file_output = opt.optarg;
			break;
		case 'b':
			bitrate = (int)(atof(opt.optarg) * 1000.0f);
			if(bitrate > 6144000 || bitrate < 32000)
			{
				fprintf(stderr, "Bitrate must be between 32 and 6144 kbps!\n");
				return 1;
			}
			break;
		case 'c':
			channel_config = atoi(opt.optarg) - 1;
			if((channel_config < 0) || (channel_config > 15))
			{
				fprintf(stderr, "Bad channel configuration. Must be between 1 and 16!\n");
				return 1;
			}
			break;
		case 'f':
			has_lfe = 1;
			break;
		case 'm':
			multi_mono = 1;
			break;
		case 'h':
			show_help = 1;
			break;
		case 'l':
			ignore_len = 1;
			break;
		case 'e':
			enc_flags = enc_flags & (~DCAENC_FLAG_BIGENDIAN);
			break;
		case 'r':
			enc_flags = enc_flags | DCAENC_FLAG_28BIT;
			break;
		case 'v':
			show_ver = 1;
			break;
		case '?':
			fprintf(stderr, "Unknown commandline option or missing argument: %s\n", argv[opt.optind-1]);
			return 1;
		}
	}
	
	// ----------------------------

	if (multi_mono) {
		any_input = 0;
		for (i=0; i<6; i++){
			if(mono_file_input[i])
			any_input = 1;
		}
	}
	else {
		if (file_input)
			any_input = 1;
		else
			any_input = 0;
	}
	if(!any_input || !file_output || bitrate < 1 || show_ver || show_help)
	{
		if(show_ver)
		{
			printf(PACKAGE_NAME "-" PACKAGE_VERSION "\n");
			printf("compiled on " __DATE__ " at " __TIME__ " using " __COMPILER__ ";\n\n");
			printf(PACKAGE_URL "\n");
			printf("http://gitorious.org/~mulder/dtsenc/mulders-dtsenc \n\n");
			printf("Copyright (c) 2008-2012 Alexander E. Patrakov <patrakov@gmail.com>\n");
			printf("This program is free software: you can redistribute it and/or modify it\n");
			printf("under the terms of the GNU Lesser General Public License.\n");
			printf("Note that this program is distributed with ABSOLUTELY NO WARRANTY.\n\n");
			printf("Portions changed by Alexei Andropov for ffdcaenc.\n\n");
			return 0;
		}
		else if(show_help)
		{
			printf("FFDCAENC --- experimental 'Coherent Acoustics' compressor.\n\n");
			printf("Usage:\n  ffdcaenc -i <input.wav> -o <output.dts> -b <bitrate_kbps>\n\n");
			printf("Optional:\n");
			printf("  -l  Ignore input length, can be useful when reading from stdin\n");
			printf("  -e  Switch output endianess to Little Endian (default is: Big-Endian)\n");
			printf("  -r  Reduced Bit Depth for DTS CD format (default is: Full Bit-Depth)\n");
			printf("  -h  Print this help screen\n");
			printf("  -c  Overwrite the channel configuration (default is: auto-selection)\n");
			printf("  -f  Add an additional LFE channel (default: used for 6-channel input)\n");
			printf("  -m  Multiple Mono input files (default: -i for multi-channel input file)\n");
			printf("       Use -0 <input.wav> -1 <input.wav> etc. (up to -5) Channels are in ITU order:\n");
			printf("       0,1,2,3,4,5 --> LF, RF, C, LFE, LS, RS\n");
			printf("       The following mono input file combinations are supported:\n");
			printf("		1.0		-2 center.wav\n");
			printf("		1.1		-2 center.wav -3 lfe-wav\n");
			printf("		2.0		-0 left.wav -1 right.wav\n");
			printf("		2.1		-0 left.wav -1 right.wav -3 lfe.wav\n");
			printf("		3.0		-0 left.wav -1 right.wav -2 center.wav\n");
			printf("		3.1		-0 left.wav -1 right.wav -2 center.wav -3 lfe.wav\n");
			printf("		4.0		-0 left.wav -1 right.wav -4 ls.wav -5 rs.wav\n");
			printf("		4.1		-0 left.wav -1 right.wav -4 ls.wav -5 rs.wav -3 lfe.wav\n");
			printf("		5.0		-0 left.wav -1 right.wav -2 center.wav -4 ls.wav -5 rs.wav\n");
			printf("		5.1		-0 left.wav -1 right.wav -2 center.wav -4 ls.wav -5 rs.wav -3 lfe.wav\n\n");
			printf("  -v  Show version info\n\n");
			printf("REMARKS:\n");
			printf("The input or output filename can be \"-\" for stdin/stdout.\n");
			printf("The bitrate is specified in kilobits per second and may be rounded up\n");
			printf("-- use floating-point values for bitrates that are not a multiple of 1 kbps.\n");
			printf("Because the encoder uses a 4-byte granularity, i.e., 32 bits per audio frame\n");
			printf("(with 512 samples/frame), the ACTUAL bitrate will always be a multiple of:\n\n");
			printf("3         kbps for     48 kHz\n");
			printf("2.75625   kbps for   44.1 kHz\n");
			printf("2         kbps for     32 kHz\n");
			printf("1.5       kbps for     24 kHz\n");
			printf("1.378125  kbps for  22.05 kHz\n");
			printf("1         kbps for     16 kHz\n");
			printf("0.75      kbps for     12 kHz\n");
			printf("0.6890625 kbps for 11.025 kHz\n");
			printf("0.5       kbps for      8 kHz\n\n");
			printf("-- NOTICE: the values 754.5 and 1509.75 AT _48kHz_ are an exception.\n\n");
			printf("* Available channel-layouts:\n\n");
			printf("  -  1: A\n");
			printf("  -  2: A, B\n");
			printf("  -  3: L, R\n");
			printf("  -  4: (L+R), (L-R)\n");
			printf("  -  5: Lt, Rt\n");
			printf("  -  6: FC, FL, FR\n");
			printf("  -  7: FL, FR, BC\n");
			printf("  -  8: FC, FL, FR, BC\n");
			printf("  -  9: FL, FR, BL, BR\n");
			printf("  - 10: FC, FL, FR, BL, BR\n");
			printf("  - 11: CL, CR, FL, FR, BL, BR (not supported)\n");
			printf("  - 12: FC, FL, FR, BL, BR, OV (not supported)\n");
			printf("  - 13: FC, BC, FL, FR, BL, BR (not supported)\n");
			printf("  - 14: CL, FC, CR, FL, FR, BL, BR (not supported)\n");
			printf("  - 15: CL, CR, FL, FR, SL1, SL2, SR1, SR2 (not supported)\n");
			printf("  - 16: CL, FC, CR, FL, FR, BL, BC, BR (not supported)\n\n");
			printf("* Valid sample rates (in kHz):\n\n  8 11.025 12 16 22.05 24 32 44.1 48\n\n");
			printf("* Transmission bitrates (in kbps):\n\n  32     56     64     96     112    128    192     224   \n  256    320    384    448    512    576    640     768   \n  960    1024   1152   1280   1344   1408   1411.2  1472  \n  1536   1920   2048   3072   3840   open   VBR     LOSSLESS\n");
			return 0;
		}
		else
		{
			printf("FFDCAENC --- experimental 'Coherent Acoustics' compressor.\n\n");
			printf("ERROR: Required arguments are missing.\nCall 'ffdcaenc -h' for more information.\n");
			return 1;
		}
	}

	if (multi_mono) {
		if ( !mono_file_input[0] && !mono_file_input[1] && !mono_file_input[2] && !mono_file_input[3] && !mono_file_input[4] && !mono_file_input[5]) {
			fprintf(stderr, "Missing Input File. -m option requires at least one input file\n");
			return 1;
		}


		if ( mono_file_input[0] )
			fprintf(stderr, "LF Source: %s\n", mono_file_input[0]);
		if ( mono_file_input[1] )
			fprintf(stderr, "RF Source: %s\n", mono_file_input[1]);
		if ( mono_file_input[2] )
			fprintf(stderr, "C Source: %s\n", mono_file_input[2]);
		if ( mono_file_input[3] )
			fprintf(stderr, "LFE Source: %s\n", mono_file_input[3]);
		if ( mono_file_input[4] )
			fprintf(stderr, "LS Source: %s\n", mono_file_input[4]);
		if ( mono_file_input[5] )
			fprintf(stderr, "RS Source: %s\n", mono_file_input[5]);
	}
	else
	fprintf(stderr, "FFDCAENC --- experimental 'Coherent Acoustics' compressor.\n\n");
	fprintf(stderr, "Input:   %s\n", file_input);
	fprintf(stderr, "Output:  %s\n", file_output);
	fprintf(stderr, "Bitrate: %u kbps\n\n", bitrate / 1000, bitrate % 1000);

	// ----------------------------

	if (!multi_mono) {

		f = wavfile_open(file_input, &error_msg, ignore_len);
		if (!f) {
			fprintf(stderr, "Could not open or parse \"%s\".\n", file_input);
			fprintf(stderr, "Error: %s!\n", error_msg);
			return 1;
		}
	
		samples_total = f->samples_left;

		if(channel_config == AUTO_SELECT)
		{
			channel_config = channel_map[f->channels - 1];
			if(f->channels == 6) has_lfe = 1;
		}

		if(has_lfe)
		{
			enc_flags = enc_flags | DCAENC_FLAG_LFE;
		}

		switch(f->channels - (has_lfe ? 1 : 0))
		{
		case 1:
			if(!(channel_config == DCAENC_CHANNELS_MONO))
			{
				fprintf(stderr, "Invalid channel configuration for input audio!\n");
				return 1;
			}
			break;
		case 2:
			if(!(channel_config == DCAENC_CHANNELS_DUAL_MONO || channel_config == DCAENC_CHANNELS_STEREO ||
				channel_config == DCAENC_CHANNELS_STEREO_SUMDIFF || channel_config == DCAENC_CHANNELS_STEREO_TOTAL))
			{
				fprintf(stderr, "Invalid channel configuration for input audio!\n");
				return 1;
			}
			break;
		case 4:
			if(!(channel_config == DCAENC_CHANNELS_3FRONT_1REAR || channel_config == DCAENC_CHANNELS_2FRONT_2REAR))
			{
				fprintf(stderr, "Invalid channel configuration for input audio!\n");
				return 1;
			}
			break;
		case 5:
			if(!(channel_config == DCAENC_CHANNELS_3FRONT_2REAR))
			{
				fprintf(stderr, "Invalid channel configuration for input audio!\n");
				return 1;
			}
			break;
		case 6:
			if(!(channel_config == DCAENC_CHANNELS_3FRONT_3REAR || channel_config == DCAENC_CHANNELS_4FRONT_2REAR ||
				 channel_config == DCAENC_CHANNELS_3FRONT_2REAR_1OV))
			{
				fprintf(stderr, "Invalid channel configuration for input audio!\n");
				return 1;
			}
			break;
		}
	}
	else { // multi_mono

		mono_channels = 0;
		has_lfe = 0;
		for (i=0; i<6; i++){
			if ( mono_file_input[i]) {
				mono_f[i] = wavfile_open(mono_file_input[i], &error_msg, ignore_len);
				if (!mono_f[i]) {
					fprintf(stderr, "Could not open or parse \"%s\".\n", mono_file_input[i]);
					fprintf(stderr, "Error: %s!\n", error_msg);
					return 1;
				}
				mono_samples_total[i] = mono_f[i]->samples_left;
				++ mono_channels;
			}
		}
		switch(mono_channels) {
		case 1:
			channel_config = DCAENC_CHANNELS_MONO;
			if (!mono_f[2]) {
				fprintf(stderr, "Mono requires -2\n");
				return 1;
			}
			an_input = 2;
			break;
		case 2:
			if ( !mono_file_input[3] ) {
				channel_config = DCAENC_CHANNELS_STEREO;
				if (!mono_f[0] ||!mono_f[1] ) {
						fprintf(stderr, "Stereo requires -0 and -1\n");
						return 1;
				}
				if ( mono_samples_total[0] != mono_samples_total[1]) {
					fprintf(stderr, "Multi mono inputs have different sample lengths\n");
					return 1;
				}
				if ( mono_f[0]->channels != 1 || mono_f[1]->channels != 1 ) {
					fprintf(stderr, "All input files must be mono (for multi-mono)\n");
					return 1;
				}
				an_input = 0;
			}
			else {
				channel_config = DCAENC_CHANNELS_MONO;
				if (!mono_f[2]) {
					fprintf(stderr, "1.1 requires -2 and -3\n");
					return 1;
				}
				if ( mono_samples_total[2] != mono_samples_total[3]) {
					fprintf(stderr, "Multi mono inputs have different sample lengths\n");
					return 1;
				}
				if ( mono_f[2]->channels != 1 || mono_f[3]->channels != 1 ) {
					fprintf(stderr, "All input files must be mono (for multi-mono)\n");
					return 1;
				}
				an_input = 2;
				has_lfe = 1;
			}
			break;
		case 3:
			if ( !mono_file_input[3] ) {
				channel_config = DCAENC_CHANNELS_3FRONT;
				if (!mono_f[0] || !mono_f[1] || !mono_f[2] ) {
					fprintf(stderr, "3.0 requires -0, -1, -2\n");
					return 1;
				}
				if ( mono_samples_total[0] != mono_samples_total[1] || mono_samples_total[0] != mono_samples_total[2]) {
					fprintf(stderr, "Multi mono inputs have different sample lengths\n");
					return 1;
				}
				if ( mono_f[0]->channels != 1 || mono_f[1]->channels != 1 || mono_f[2]->channels != 1) {
					fprintf(stderr, "All input files must be mono (for multi-mono)\n");
					return 1;
				}
				an_input = 0;
			}
			else {
				channel_config = DCAENC_CHANNELS_STEREO;
				if (!mono_f[0] ||!mono_f[1] ) {
						fprintf(stderr, "2.1 requires -0, -1 and -2\n");
						return 1;
				}
				if ( mono_samples_total[0] != mono_samples_total[1] || mono_samples_total[0] != mono_samples_total[3]) {
					fprintf(stderr, "Multi mono inputs have different sample lengths\n");
					return 1;
				}
				if ( mono_f[0]->channels != 1 || mono_f[1]->channels != 1 || mono_f[3]->channels != 1) {
					fprintf(stderr, "All input files must be mono (for multi-mono)\n");
					return 1;
				}
				an_input = 0;
				has_lfe = 1;
			}
			break;
		case 4:
			if ( !mono_file_input[3] ) {
				channel_config = DCAENC_CHANNELS_2FRONT_2REAR;
				if (!mono_f[0] || !mono_f[1] || !mono_f[4] || !mono_f[5]) {
						fprintf(stderr, "4.0 requires -0, -1, -4, -5\n");
						return 1;
					}
				if ( mono_samples_total[0] != mono_samples_total[1] || mono_samples_total[0] != mono_samples_total[4] || mono_samples_total[0] != mono_samples_total[5]) {
					fprintf(stderr, "Multi mono inputs have different sample lengths\n");
					return 1;
				}
				if ( mono_f[0]->channels != 1 || mono_f[1]->channels != 1 || mono_f[4]->channels != 1 || mono_f[5]->channels != 1 ) {
					fprintf(stderr, "All input files must be mono (for multi-mono)\n");
					return 1;
				}
				an_input = 0;
			}
			else {
				channel_config = DCAENC_CHANNELS_3FRONT;
				if (!mono_f[0] || !mono_f[1] || !mono_f[2] ) {
					fprintf(stderr, "3.1 requires -0, -1, -2, and -3\n");
					return 1;
				}
				if ( mono_samples_total[0] != mono_samples_total[1] || mono_samples_total[0] != mono_samples_total[2] || mono_samples_total[0] != mono_samples_total[3]) {
					fprintf(stderr, "Multi mono inputs have different sample lengths\n");
					return 1;
				}
				if ( mono_f[0]->channels != 1 || mono_f[1]->channels != 1 || mono_f[2]->channels != 1 || mono_f[3]->channels != 1) {
					fprintf(stderr, "All input files must be mono (for multi-mono)\n");
					return 1;
				}
				an_input = 0;
				has_lfe = 1;
			}
			break;
		case 5:
			if ( !mono_file_input[3] ) {
				channel_config = DCAENC_CHANNELS_3FRONT_2REAR;
				if (!mono_f[0] || !mono_f[1] || !mono_f[2] || !mono_f[4] || !mono_f[5]) {
					fprintf(stderr, "5.0 requires -0, -1, -2, -4, -5\n");
					return 1;
				}
				if ( mono_samples_total[0] != mono_samples_total[1] || mono_samples_total[0] != mono_samples_total[2] || mono_samples_total[0] != mono_samples_total[4] || mono_samples_total[0] != mono_samples_total[5]) {
					fprintf(stderr, "Multi mono inputs have different sample lengths\n");
					return 1;
				}
				if ( mono_f[0]->channels != 1 || mono_f[1]->channels != 1 || mono_f[2]->channels != 1 || mono_f[4]->channels != 1 || mono_f[5]->channels != 1 ) {
					fprintf(stderr, "All input files must be mono (for multi-mono)\n");
					return 1;
				}
			}
			else {
				channel_config = DCAENC_CHANNELS_2FRONT_2REAR;
				if (!mono_f[0] || !mono_f[1] || !mono_f[3] || !mono_f[4] || !mono_f[5]) {
						fprintf(stderr, "4.1 requires -0, -1, -3, -4, -5\n");
						return 1;
				}
				if ( mono_samples_total[0] != mono_samples_total[1] || mono_samples_total[0] != mono_samples_total[3] || mono_samples_total[0] != mono_samples_total[4] || mono_samples_total[0] != mono_samples_total[5]) {
					fprintf(stderr, "Multi mono inputs have different sample lengths\n");
					return 1;
				}
				if ( mono_f[0]->channels != 1 || mono_f[1]->channels != 1 || mono_f[3]->channels != 1|| mono_f[4]->channels != 1 || mono_f[5]->channels != 1 ) {
					fprintf(stderr, "All input files must be mono (for multi-mono)\n");
					return 1;
				}
				has_lfe = 1;
			}
			an_input = 0;
			break;
		case 6:
			channel_config = DCAENC_CHANNELS_3FRONT_2REAR;
			if ( mono_samples_total[0] != mono_samples_total[1] || mono_samples_total[0] != mono_samples_total[2]  || mono_samples_total[0] != mono_samples_total[3] || mono_samples_total[0] != mono_samples_total[4] || mono_samples_total[0] != mono_samples_total[5]) {
				fprintf(stderr, "Multi mono inputs have different sample lengths\n");
				return 1;
			}
			if ( mono_f[0]->channels != 1 || mono_f[1]->channels != 1 || mono_f[2]->channels != 1 || mono_f[3]->channels != 1 || mono_f[4]->channels != 1 || mono_f[5]->channels != 1 ) {
				fprintf(stderr, "All input files must be mono (for multi-mono)\n");
				return 1;
			}
			has_lfe = 1;
			break;
			an_input = 0;
		}	

		if(has_lfe)
		{
			enc_flags = enc_flags | DCAENC_FLAG_LFE;
		}


	}

	// ----------------------------

	c = (multi_mono ? dcaenc_create(mono_f[an_input]->sample_rate, channel_config, bitrate, enc_flags) : dcaenc_create(f->sample_rate, channel_config, bitrate, enc_flags));
	
	if (!c) {
		fprintf(stderr, "ERROR: Invalid bitrate or sample rate.\n");
		return 1;
	}


	if((((size_t)(dcaenc_output_size(c))) > BUFFSIZE_OUT) || (((size_t)(dcaenc_input_size(c))) > BUFFSIZE_SPL))
	{
		fprintf(stderr, "Internal error, buffers are too small!\n", file_output);
		return 1;
	}
	
	outfile = strcmp(file_output, "-") ? fopen(file_output, "wb") : stdout;
	if(!outfile) {
		fprintf(stderr, "Could not open \"%s\" for writing!\n", file_output);
		return 1;
	}
	
	fflush(stdout);
	fflush(stderr);
	
	// ----------------------------

	counter = 0;
	samples_read_total = 0;
	status_idx = 0;
	
	while(samples_read = ( multi_mono ? multi_mono_wavfile_read_s32(mono_f[an_input], mono_f[1], mono_f[2], mono_f[3], mono_f[4], mono_f[5], data, BUFFSIZE_SPL, channel_config, has_lfe) : wavfile_read_s32(f, data, BUFFSIZE_SPL)))
	{
		samples_read_total += samples_read;
		wrote = dcaenc_convert_s32(c, data, output);
		fwrite(output, 1, wrote, outfile);
		
		if(counter == 0)
		{
			current_pos = samples_read_total / (multi_mono ? mono_f[an_input]->sample_rate :f->sample_rate) ;
			
			if((samples_total > 0) && (samples_total < UNKNOWN_SIZE))
			{
				percent_done = ((double)(samples_total - (multi_mono ? mono_f[an_input]->samples_left :f->samples_left))) / ((double)(samples_total));
				fprintf(stderr, "Encoding... %u:%02u [%3.1f%%]\r", current_pos / 60, current_pos % 60, percent_done * 100.0);
				fflush(stderr);
			}
			else
			{
				fprintf(stderr, "Encoding... %u:%02u [%c]\r", current_pos / 60, current_pos % 60, status[(status_idx = (status_idx+1) % 4)]);
				fflush(stderr);
			}
		}
		
		counter = (counter+1) % 125;
	}
	
	fprintf(stderr, "Encoding... %u:%02u [%3.1f%%]\n", (samples_read_total / (multi_mono ? mono_f[an_input]->sample_rate :f->sample_rate)) / 60, (samples_read_total / (multi_mono ? mono_f[an_input]->sample_rate :f->sample_rate)) % 60, 100.0);
	fflush(stderr);

	wrote = dcaenc_destroy(c, output);
	fwrite(output, 1, wrote, outfile);
	if(outfile != stdout)
	{
		fclose(outfile);
	}
	if (!multi_mono)
		wavfile_close(f);
	else {
		if ( channel_config != DCAENC_CHANNELS_MONO ) {
			wavfile_close(mono_f[0]);
			wavfile_close(mono_f[1]);
		}
		if ( channel_config == DCAENC_CHANNELS_MONO || channel_config == DCAENC_CHANNELS_3FRONT_2REAR || channel_config == DCAENC_CHANNELS_3FRONT)
			wavfile_close(mono_f[2]);
		if(has_lfe)
			wavfile_close(mono_f[3]);
		if ( channel_config == DCAENC_CHANNELS_2FRONT_2REAR || channel_config == DCAENC_CHANNELS_3FRONT_2REAR )  {
			wavfile_close(mono_f[4]);
			wavfile_close(mono_f[5]);
		}
	}

	fprintf(stderr, "Done.\n");
	return 0;
}


int32_t main( int32_t argc, char **argv )
{
	return dcaenc_main(argc, argv);
}

